File: static/js/hipoteca.js
/* Motor del simulador de vivienda.
 *
 * Todo son funciones puras: entran números, salen números. Viven aquí y no en
 * la plantilla para poder probarlas de verdad (y para que la plantilla se lea).
 * Los porcentajes de impuestos y los valores por defecto los pone el servidor
 * (finanzas/hipoteca.py) y llegan como datos, para que no haya dos verdades.
 */
(function (global) {
'use strict';

// ── Préstamo francés ────────────────────────────────────────────────────────

function cuotaMensual(capital, tipoAnual, años) {
    if (!(capital > 0) || !(años > 0)) return 0;
    const n = años * 12;
    if (!tipoAnual) return capital / n;
    const r = tipoAnual / 100 / 12;
    return capital * r * Math.pow(1 + r, n) / (Math.pow(1 + r, n) - 1);
}

function capitalMaximo(cuotaMax, tipoAnual, años) {
    if (!(cuotaMax > 0) || !(años > 0)) return 0;
    const n = años * 12;
    if (!tipoAnual) return cuotaMax * n;
    const r = tipoAnual / 100 / 12;
    return cuotaMax * (Math.pow(1 + r, n) - 1) / (r * Math.pow(1 + r, n));
}

/* Tipo que se aplica cada año según la modalidad.
 *   fijo     → el mismo siempre
 *   variable → euríbor + diferencial desde el primer día
 *   mixto    → un tipo fijo los primeros `añosFijos`, luego euríbor + diferencial
 * `euriborDelta` desplaza el euríbor: es la palanca del test de estrés. */
function tipoEnAño(cfg, año) {
    const euribor = (cfg.euribor || 0) + (cfg.euriborDelta || 0);
    if (cfg.modalidad === 'fijo') return cfg.tipoFijo;
    if (cfg.modalidad === 'variable') return euribor + cfg.diferencial;
    return año < cfg.añosFijos ? cfg.tipoFijo : euribor + cfg.diferencial;
}

/* Cuadro de amortización año a año.
 *
 * En variable y mixto la cuota se recalcula cuando cambia el tipo, sobre el
 * capital que queda y el plazo que resta, que es como funciona una revisión.
 * `amortizacionAnual` es dinero extra que se mete cada año: si `modo` es
 * 'plazo' acorta el préstamo (y la cuota no baja); si es 'cuota', la recalcula
 * sobre el plazo original. */
function cuadroAmortizacion(capital, cfg, años, amortizacionAnual, modo) {
    amortizacionAnual = amortizacionAnual || 0;
    const filas = [];
    let pendiente = capital;
    let plazoRestante = años;
    let tipoAnterior = null;
    let cuota = 0;
    let interesesAcum = 0;

    for (let a = 0; a < años && pendiente > 0.01; a++) {
        const tipo = tipoEnAño(cfg, a);
        if (tipo !== tipoAnterior || cuota === 0) {
            cuota = cuotaMensual(pendiente, tipo, plazoRestante);
            tipoAnterior = tipo;
        }
        const r = tipo / 100 / 12;
        let interesAño = 0, capitalAño = 0;

        for (let m = 0; m < 12 && pendiente > 0.01; m++) {
            const interes = pendiente * r;
            let amortiza = Math.min(cuota - interes, pendiente);
            if (amortiza < 0) amortiza = 0;
            pendiente -= amortiza;
            interesAño += interes;
            capitalAño += amortiza;
        }

        if (amortizacionAnual > 0 && pendiente > 0.01) {
            const extra = Math.min(amortizacionAnual, pendiente);
            pendiente -= extra;
            capitalAño += extra;
            if (modo === 'cuota') {
                cuota = cuotaMensual(pendiente, tipo, años - a - 1);
                tipoAnterior = null;   // fuerza recálculo el año siguiente
            }
        }

        interesesAcum += interesAño;
        filas.push({
            año: a + 1, tipo: tipo, cuota: cuota,
            intereses: interesAño, capital: capitalAño,
            pendiente: Math.max(0, pendiente), interesesAcum: interesesAcum,
        });
        plazoRestante = Math.max(1, plazoRestante - 1);
    }
    return filas;
}

function totalIntereses(capital, cfg, años, amortizacionAnual, modo) {
    const filas = cuadroAmortizacion(capital, cfg, años, amortizacionAnual, modo);
    return filas.length ? filas[filas.length - 1].interesesAcum : 0;
}

/* La cuota que hay que poder pagar, no la del primer año: en variable y mixto
 * la de partida es la más baja del préstamo, y decidir con ella es engañarse. */
function cuotaMaxima(capital, cfg, años) {
    const filas = cuadroAmortizacion(capital, cfg, años, 0, 'plazo');
    return filas.reduce(function (m, f) { return Math.max(m, f.cuota); }, 0);
}

// ── Gastos de compra ────────────────────────────────────────────────────────

function gastosCompra(precio, ccaa, obraNueva, joven, D) {
    if (!(precio > 0)) {
        return {impuestos: 0, notaria: 0, registro: 0, gestoria: 0, tasacion: 0,
                total: 0, totalPct: 0, impuestoNombre: '—', impuestoPct: 0};
    }
    let impuestoPct, impuestoNombre;
    if (obraNueva) {
        impuestoPct = D.iva_obra_nueva + ccaa.ajd;
        impuestoNombre = 'IVA ' + D.iva_obra_nueva + '% + AJD ' + ccaa.ajd + '%';
    } else {
        impuestoPct = (joven && ccaa.itp_joven !== null) ? ccaa.itp_joven : ccaa.itp;
        impuestoNombre = 'ITP ' + impuestoPct + '%';
    }
    const entre = function (v, lo, hi) { return Math.max(lo, Math.min(hi, v)); };
    const impuestos = precio * impuestoPct / 100;
    const notaria   = entre(precio * D.notaria_pct / 100, D.notaria_min, D.notaria_max);
    const registro  = entre(precio * D.registro_pct / 100, D.registro_min, D.registro_max);
    const total = impuestos + notaria + registro + D.gestoria + D.tasacion;
    return {
        impuestos: impuestos, notaria: notaria, registro: registro,
        gestoria: D.gestoria, tasacion: D.tasacion,
        total: total, totalPct: total / precio * 100,
        impuestoNombre: impuestoNombre, impuestoPct: impuestoPct,
    };
}

// ── Coste de tener la vivienda ──────────────────────────────────────────────

function costeRecurrente(valor, p) {
    const mantenimiento = valor * p.mantenimientoPct / 100 / 12;
    const ibi = valor * p.ibiPct / 100 / 12;
    const seguro = p.seguroAnual / 12;
    return {
        mantenimiento: mantenimiento, ibi: ibi, seguro: seguro,
        comunidad: p.comunidadMensual,
        total: mantenimiento + ibi + seguro + p.comunidadMensual,
    };
}

// ── Capacidad de compra ─────────────────────────────────────────────────────

/* Precio máximo alcanzable con un capital y una cuota tope.
 *
 * La cuota tope manda sobre el capital que se puede pedir; el capital propio
 * tiene que cubrir la entrada mínima MÁS los gastos de compra. Se resuelve
 * iterando porque los gastos dependen del precio (y el impuesto es un % de él). */
function precioMaximo(capitalDisponible, cuotaTope, cfg, años, entradaMinPct, ccaa, obraNueva, joven, D) {
    const hipMax = capitalMaximo(cuotaTope, tipoEnAño(cfg, 0), años);
    let precio = capitalDisponible + hipMax;
    for (let i = 0; i < 40; i++) {
        const g = gastosCompra(precio, ccaa, obraNueva, joven, D);
        // Con gastos g, el capital que queda para la entrada es este:
        const paraEntrada = Math.max(0, capitalDisponible - g.total);
        // Precio que sale de juntar entrada + hipoteca, respetando la entrada mínima
        const porCapital = paraEntrada + hipMax;
        const porEntrada = entradaMinPct > 0 ? paraEntrada / (entradaMinPct / 100) : porCapital;
        const nuevo = Math.max(0, Math.min(porCapital, porEntrada));
        if (Math.abs(nuevo - precio) < 1) { precio = nuevo; break; }
        precio = nuevo;
    }
    return Math.max(0, precio);
}

// ── Comprar vs alquilar ─────────────────────────────────────────────────────

/* Comprar frente a seguir de alquiler, año a año.
 *
 * Los dos parten del MISMO dinero (la entrada más los gastos de compra) y del
 * mismo sueldo, así que se pueden comparar sin trampa:
 *
 *   · El comprador mete ese dinero en la casa y cada año paga cuota y costes
 *     de tenerla. Su patrimonio es lo que vale la casa menos lo que debe.
 *   · El inquilino conserva ese dinero invertido y, además, mete en la cartera
 *     la diferencia entre lo que le costaría comprar y lo que paga de alquiler
 *     (si el alquiler es más caro, la cartera mengua). Su patrimonio es la
 *     cartera.
 *
 * Esa segunda parte es la que se suele olvidar, y sin ella la comparación está
 * trucada a favor de comprar.
 */
function compararComprarAlquilar(o) {
    const filas = [];
    const cuadro = cuadroAmortizacion(o.hipoteca, o.cfg, o.años, 0, 'plazo');

    let alquilerAnual = o.alquilerMensual * 12;
    let cartera = o.capitalEntrada + o.gastosCompra;   // lo que el inquilino NO gasta
    let añoEquilibrio = null;

    for (let a = 0; a < o.años; a++) {
        const fila = cuadro[a];
        if (!fila) break;
        const valorVivienda = o.precio * Math.pow(1 + o.revalorizacionPct / 100, a + 1);
        const recurrenteAnual = costeRecurrente(valorVivienda, o.recurrente).total * 12;
        const desembolsoComprador = fila.cuota * 12 + recurrenteAnual;

        cartera *= (1 + o.rentabilidadPct / 100);
        cartera += (desembolsoComprador - alquilerAnual);

        const netoComprar = valorVivienda - fila.pendiente;
        const netoAlquilar = cartera;

        if (añoEquilibrio === null && netoComprar >= netoAlquilar) añoEquilibrio = a + 1;

        filas.push({
            año: a + 1, netoComprar: netoComprar, netoAlquilar: netoAlquilar,
            alquilerAnual: alquilerAnual, valorVivienda: valorVivienda,
            interesesAño: fila.intereses, recurrenteAño: recurrenteAnual,
        });
        alquilerAnual *= (1 + o.subidaAlquilerPct / 100);
    }
    return {filas: filas, añoEquilibrio: añoEquilibrio};
}

// ── ¿Cuándo llego? ──────────────────────────────────────────────────────────

function mesesHastaReunir(falta, ahorroMensual) {
    if (falta <= 0) return 0;
    if (!(ahorroMensual > 0)) return null;
    return Math.ceil(falta / ahorroMensual);
}

global.Hipoteca = {
    cuotaMensual: cuotaMensual,
    capitalMaximo: capitalMaximo,
    tipoEnAño: tipoEnAño,
    cuadroAmortizacion: cuadroAmortizacion,
    totalIntereses: totalIntereses,
    cuotaMaxima: cuotaMaxima,
    gastosCompra: gastosCompra,
    costeRecurrente: costeRecurrente,
    precioMaximo: precioMaximo,
    compararComprarAlquilar: compararComprarAlquilar,
    mesesHastaReunir: mesesHastaReunir,
};

})(typeof window !== 'undefined' ? window : globalThis);

File: finanzas/hipoteca.py
"""Datos y cálculos de la compra de vivienda.

Aquí vive lo que NO depende de la interacción: la tabla de impuestos por
comunidad autónoma y el desglose de los gastos de compra. El cálculo de cuotas
y escenarios vive en `static/js/hipoteca.js`, porque tiene que recalcularse a
cada movimiento de un slider; este módulo le pasa la tabla y comparte con él
los mismos porcentajes, de modo que no hay dos verdades.

AVISO sobre los impuestos: los tipos cambian con cada presupuesto autonómico y
las bonificaciones tienen letra pequeña (límite de precio, empadronamiento,
plazo para escriturar...). Lo de aquí es una ESTIMACIÓN de partida; el
simulador deja ajustar el porcentaje final a mano, que es lo que hay que hacer
antes de firmar nada.
"""

from decimal import Decimal

# Tipo general del ITP (vivienda de segunda mano) y tipo reducido habitual para
# menores de 35 años comprando vivienda habitual, cuando la comunidad lo tiene.
# `ajd` es el impuesto de Actos Jurídicos Documentados, que es el que se paga
# en obra nueva (junto al IVA del 10 %).
CCAA = [
    # código, nombre, itp general %, itp joven %, ajd %
    ('AN', 'Andalucía',            7.0,  3.5,  1.2),
    ('AR', 'Aragón',               8.0,  4.0,  1.5),
    ('AS', 'Asturias',             8.0,  3.0,  1.2),
    ('IB', 'Baleares',             8.0,  5.0,  1.5),
    ('CN', 'Canarias',             6.5,  5.0,  0.75),
    ('CB', 'Cantabria',            9.0,  5.0,  1.5),
    ('CM', 'Castilla-La Mancha',   9.0,  6.0,  1.5),
    ('CL', 'Castilla y León',      8.0,  5.0,  1.5),
    ('CT', 'Cataluña',            10.0,  5.0,  1.5),
    ('EX', 'Extremadura',          8.0,  6.0,  1.5),
    ('GA', 'Galicia',              9.0,  3.0,  1.5),
    ('MD', 'Madrid',               6.0,  6.0,  0.75),
    ('MC', 'Murcia',               8.0,  3.0,  1.5),
    ('NC', 'Navarra',              6.0,  5.0,  0.5),
    ('PV', 'País Vasco',           4.0,  2.5,  0.5),
    ('RI', 'La Rioja',             7.0,  5.0,  1.0),
    ('VC', 'Comunidad Valenciana', 10.0, 8.0,  1.5),
]

IVA_OBRA_NUEVA = 10.0

# Gastos fijos de la operación. Son estimaciones de mercado: un porcentaje con
# suelo y techo, que es como se comportan los aranceles.
NOTARIA_PCT, NOTARIA_MIN, NOTARIA_MAX = 0.35, 600, 1400
REGISTRO_PCT, REGISTRO_MIN, REGISTRO_MAX = 0.20, 400, 800
GESTORIA = 350
TASACION = 350

# Coste recurrente de tener la vivienda, aparte de la cuota. Son los valores de
# partida del simulador; todos se pueden ajustar.
MANTENIMIENTO_PCT_ANUAL = 1.0    # % del valor del inmueble al año
SEGURO_HOGAR_ANUAL = 300
IBI_PCT_ANUAL = 0.5              # % del valor, orientativo (varía por municipio)
COMUNIDAD_MENSUAL = 60


def tabla_ccaa():
    """La tabla en el formato que consume el JS del simulador."""
    return [
        {'codigo': c, 'nombre': n, 'itp': itp, 'itp_joven': joven, 'ajd': ajd}
        for c, n, itp, joven, ajd in CCAA
    ]


def _entre(valor, minimo, maximo):
    return max(minimo, min(maximo, valor))


def gastos_compra(precio, ccaa='MD', obra_nueva=False, joven=False):
    """Desglose de lo que cuesta comprar, aparte del precio.

    Devuelve importes en euros y el total. En obra nueva se paga IVA + AJD; en
    segunda mano, ITP (con el tipo reducido si aplica la bonificación joven)."""
    precio = float(precio or 0)
    fila = next((f for f in CCAA if f[0] == ccaa), None)
    if fila is None:
        fila = next(f for f in CCAA if f[0] == 'MD')
    _, _, itp_general, itp_joven, ajd_pct = fila

    if obra_nueva:
        impuesto_pct = IVA_OBRA_NUEVA + ajd_pct
        impuesto_nombre = f'IVA {IVA_OBRA_NUEVA:g}% + AJD {ajd_pct:g}%'
    else:
        impuesto_pct = itp_joven if (joven and itp_joven is not None) else itp_general
        impuesto_nombre = f'ITP {impuesto_pct:g}%'

    impuestos = precio * impuesto_pct / 100
    notaria = _entre(precio * NOTARIA_PCT / 100, NOTARIA_MIN, NOTARIA_MAX) if precio else 0
    registro = _entre(precio * REGISTRO_PCT / 100, REGISTRO_MIN, REGISTRO_MAX) if precio else 0
    gestoria = GESTORIA if precio else 0
    tasacion = TASACION if precio else 0

    total = impuestos + notaria + registro + gestoria + tasacion
    return {
        'impuestos': round(impuestos, 2),
        'impuesto_pct': impuesto_pct,
        'impuesto_nombre': impuesto_nombre,
        'notaria': round(notaria, 2),
        'registro': round(registro, 2),
        'gestoria': gestoria,
        'tasacion': tasacion,
        'total': round(total, 2),
        'total_pct': round(total / precio * 100, 2) if precio else 0,
    }


def cuota_mensual(capital, tipo_anual, años):
    """Cuota de un préstamo francés. Es la misma fórmula que usa el JS; está
    aquí para poder comprobarla con tests."""
    capital = float(capital or 0)
    años = int(años or 0)
    if capital <= 0 or años <= 0:
        return 0.0
    n = años * 12
    if not tipo_anual:
        return capital / n
    r = float(tipo_anual) / 100 / 12
    return capital * r * (1 + r) ** n / ((1 + r) ** n - 1)


def capital_maximo(cuota_max, tipo_anual, años):
    """Cuánto capital soporta una cuota dada: la inversa de `cuota_mensual`."""
    cuota_max = float(cuota_max or 0)
    años = int(años or 0)
    if cuota_max <= 0 or años <= 0:
        return 0.0
    n = años * 12
    if not tipo_anual:
        return cuota_max * n
    r = float(tipo_anual) / 100 / 12
    return cuota_max * ((1 + r) ** n - 1) / (r * (1 + r) ** n)


def coste_recurrente_mensual(valor_vivienda, mantenimiento_pct=None, ibi_pct=None,
                             seguro_anual=None, comunidad_mensual=None):
    """Lo que cuesta TENER la vivienda cada mes, sin contar la hipoteca.

    Es la diferencia entre poder pagar la cuota y poder permitirse la casa."""
    valor = float(valor_vivienda or 0)
    mantenimiento_pct = MANTENIMIENTO_PCT_ANUAL if mantenimiento_pct is None else mantenimiento_pct
    ibi_pct = IBI_PCT_ANUAL if ibi_pct is None else ibi_pct
    seguro_anual = SEGURO_HOGAR_ANUAL if seguro_anual is None else seguro_anual
    comunidad_mensual = COMUNIDAD_MENSUAL if comunidad_mensual is None else comunidad_mensual

    mantenimiento = valor * float(mantenimiento_pct) / 100 / 12
    ibi = valor * float(ibi_pct) / 100 / 12
    seguro = float(seguro_anual) / 12
    return {
        'mantenimiento': round(mantenimiento, 2),
        'ibi': round(ibi, 2),
        'seguro': round(seguro, 2),
        'comunidad': round(float(comunidad_mensual), 2),
        'total': round(mantenimiento + ibi + seguro + float(comunidad_mensual), 2),
    }


def defaults_para_js():
    """Valores de partida que el simulador enseña antes de que toques nada."""
    return {
        'iva_obra_nueva': IVA_OBRA_NUEVA,
        'notaria_pct': NOTARIA_PCT, 'notaria_min': NOTARIA_MIN, 'notaria_max': NOTARIA_MAX,
        'registro_pct': REGISTRO_PCT, 'registro_min': REGISTRO_MIN, 'registro_max': REGISTRO_MAX,
        'gestoria': GESTORIA, 'tasacion': TASACION,
        'mantenimiento_pct': MANTENIMIENTO_PCT_ANUAL,
        'ibi_pct': IBI_PCT_ANUAL,
        'seguro_anual': SEGURO_HOGAR_ANUAL,
        'comunidad_mensual': COMUNIDAD_MENSUAL,
    }
